env/file_system_tracer.h

```diff
@@ -131,26 +131,30 @@ class FSSequentialFileTracingWrapper : public FSSequentialFileWrapper {
 // FSSequentialFileTracingWrapper when tracing is disabled.
 class FSSequentialFilePtr {
```

Comment:
Can you explain why we need this class at all?  Why not have a the caller decide whether to create a "simple" FSSequentialFile or one that is a FSSequentialFileTracingWrapper?  

Changing to using this class (instead of the underlying FSSequentialFile) creates an additional layer of wrapping and causes a bunch of methods to be changed for no material gain.

Reply:
When user enables IOTracing then  we call FSSequentialFileTracingWrapper. It creates trace log and call appropriate function to dump that trace log and underlying file system. But in the middle if user wants to disable the tracing. We have no way of finding if tracing is disabled or not in FileSystem level because its pointer has been created to point to FSSequentialFileTracingWrapper. So even if the tracing is disabled it does extra work of creating trace log and passing additional context as part of trace log(like thread id, or other info  not passed yet).  So we want to bypass that.

Reply:
What does "in the middle" mean?  Once an object is created as traced or not, you are always still going the same direction.

This changeset is very large to solve a very small, localized problem.  

It is much simpler to to create the right class (if given an IOTracer, create a TracingWrapper, if not create a non-wrapped one).  Inside the TracingWrapper, you can decide whether or not to trace based on the current IOTracer settings.  Inside of the TracingWrapper you can handle all of the cases required (like when to create a trace log, etc) based on the settings.  The log may be created when the Wrapper is first created or it may be created later if the settings changed (if that is allowed).  You should be able to have the TracingWrapper be almost no additional code or work if tracing is disabled.

Reply:
> What does "in the middle" mean?

@akankshamahajan15 I would also like to understand what you mean by that please.

Reply:
Sorry for not being clear in explaining. What I meant was after few DB operations if user want to end the tracing and perform remaining operations, and later on wants to enable the tracing again for few operations.                                        I added a wrapper that overloaded operator -> because in FSSequentialFile, FSWritableFileetc, adding an extra wrapper (virtual functions) and passing calls from TracingWrapper to FileSystem might be costly (tracing is not enabled) . So in order to avoid extra cost when tracing is not enabled and bypass calls from tracing wrapper to underlying File system when tracing is disabled, this new class(FSSequentialFilePtr) is added.

Reply:
I am still not sure I understand how having this class helps this situation at all.  In order to enable or use tracing, the user must have supplied the io_tracer.  If one is provided and tracing is on, give them back a wrapped class.  Otherwise give them back a raw class.  Or for finer grain, give them a wrapped class if a tracer is provided and for every call of that wrapped class decide whether or not to do the tracing.

Changing all of the APIs to take a different object rather than hiding the complexities inside of a single method seems like a lot of extra complexity and code changes and against some of the basic OO principles.

Reply:
The ```FSSequentialFile``` pointer is stored inside ```SequenceFileReader```, so we need to pass ```io_tracer_``` to the constructor in order to allocate the tracing wrapper and store it there.